### PR TITLE
Be less aggressive about parsing out accessors in an accessor list.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/4815

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -3411,12 +3416,53 @@ private PostSkipAction SkipBadAccessorListTokens(ref SyntaxToken openBrace, Synt
...
                 || this.CurrentToken.Kind == SyntaxKind.OpenBraceToken  // for accessor blocks w/ missing keyword
-                || this.CurrentToken.Kind == SyntaxKind.SemicolonToken; // for empty body accessors w/ missing keyword
+                || this.CurrentToken.Kind == SyntaxKind.SemicolonToken // for empty body accessors w/ missing keyword
+                )
+            {
+                return true;
+            }
+
+
+            return IsPossibleAccessorModifier();
```
Comment (Eyas) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: This should be exactly equivalent to `return (a || b || c || ... || IsPossibleAccessorModifier() );`

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs [hunk 2]
```diff
@@ -8350,5 +8350,22 @@ class C
...
+{
+    public string Property1 { get; set; }
+}
+
+class Class2
+{
+    public string Property { get { return this.Source.$$
+    public Class1 Source { get; set; }
+}";
+            VerifyAnyItemExists(markup);
```
Comment (Eyas) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs]: Can we achieve the stronger guarantee that a property of `Class1` exists on `Source.`? I.e. to verify that `Property1` exists, instead of `AnyItem`?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Done!

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs]: Done!

**Comment by CyrusNajmabadi** [comment]
retest this please

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs [hunk 3]
```diff
@@ -2305,12 +2305,8 @@ public static int Main()
...
     //         get return 1;
     Diagnostic(ErrorCode.ERR_SemiOrLBraceExpected, "return"),
-    // (9,15): error CS1014: A get or set accessor expected
-    // public static int Main()
-    Diagnostic(ErrorCode.ERR_GetOrSetExpected, "int"),
-    // (9,15): error CS1513: } expected
-    // public static int Main()
-    Diagnostic(ErrorCode.ERR_RbraceExpected, "int"));
+    // (8,2): error CS1513: } expected
+    Diagnostic(ErrorCode.ERR_RbraceExpected, ""));
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs]: Error recovert is better here.  We did not try to consume "public static int Main" as an accessor.

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParsingErrorRecoveryTests.cs [hunk 4]
```diff
@@ -5494,7 +5494,7 @@ public void TestModifiedMemberAfterPropertyGetBrace()
             Assert.False(pd.AccessorList.OpenBraceToken.IsMissing);
             Assert.NotNull(pd.AccessorList.CloseBraceToken);
             Assert.True(pd.AccessorList.CloseBraceToken.IsMissing);
-            Assert.Equal(2, pd.AccessorList.Accessors.Count);
+            Assert.Equal(1, pd.AccessorList.Accessors.Count);
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParsingErrorRecoveryTests.cs]: Error recovery got better here as well.  We understand that "public class d" is not an accessor.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 5]
```diff
@@ -3413,10 +3418,52 @@ private bool IsPossibleAccessor()
...
+        {
+            // We only want to accept a modifier as the start of an accessor if the modifiers are
+            // actually followed by "get/set/add/remove".  Otherwise, we might thing think we're 
+            // starting an accessor when we're actually starting a normal class member.  For example:
+            //
+            //      class C {
+            //          public int Prop { get { this.
+            //          private DateTime x;
+            //
+            // We don't want to think of the "private" in "private int prop" as starting an accessor
```
Comment (Pilchie) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: You mean in `private DateTime x`.

File: src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs [hunk 6]
```diff
@@ -2547,7 +2547,7 @@ static void Main(string[] args)
             var tree = compilation.SyntaxTrees.Single();
             var model = compilation.GetSemanticModel(tree);
 
-            var expr = tree.GetRoot().DescendantNodes().OfType<ParameterSyntax>().Single();
+            var expr = tree.GetRoot().DescendantNodes().OfType<ParameterSyntax>().First();
```
Comment (Pilchie) [inline src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs]: This seems unrelated.

**Comment by Pilchie** [comment]
LGTM, but someone more into the parser should probably also sign off.

**Comment by CyrusNajmabadi** [comment]
Reviewers: @gafter @amcasey

File: src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs (hunk 6 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoTests.cs]: it is related.  We now parse hte above code much better than before.  This means we now see multiple parameter (the "string[] args" one and the "x" in "x =>" one).  So "Single" was causing this test to crash.

**Comment by amcasey** [comment]
FYI @VSadov @AlekseyTs @jaredpar @agocke

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 7]
```diff
@@ -3413,10 +3418,52 @@ private bool IsPossibleAccessor()
...
+            // starting an accessor when we're actually starting a normal class member.  For example:
+            //
+            //      class C {
+            //          public int Prop { get { this.
+            //          private DateTime x;
+            //
+            // We don't want to think of the "private" in "private DateTime x" as starting an accessor
+            // here.  If we do, we'll get totally thrown off in parsing the remainder and that will
+            // throw off the rest of the features that depend on a good syntax tree.
+            if (IsPossibleModifier())
```
Comment (amcasey) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Seems like you only want to consider accessibility modifiers.

Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: I thought about that.  But i could see someone writing "abstract get".  We want to correctly think of that as an accessor so we can tell them that it's an invalid modifier.  I'll add a comment to this effect.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 8]
```diff
@@ -3413,10 +3418,52 @@ private bool IsPossibleAccessor()
...
+            {
+                var peekIndex = 1;
+                while (IsPossibleModifier(this.PeekToken(peekIndex)))
```
Comment (amcasey) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: This will never go out of bounds because there will always be an EOF token and the condition will fail for that token?

Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Correct.

**Comment by amcasey** [comment]
Can you please add a parser regression test?

**Comment by amcasey** [comment]
LGTM, but you should probably get at least one compiler sign-off.

**Comment by CyrusNajmabadi** [comment]
> Can you please add a parser regression test?

I was considering that. But this seems covered already by compiler syntax tests (including the ones I had to update to now match the better behavior).

**Comment by amcasey** [comment]
I'm fine with updating one of the existing tests to show the new tree structure.  I don't think the new diagnostics capture the behavior on their own.

**Comment by CyrusNajmabadi** [comment]
@amcasey Will do!

**Comment by CyrusNajmabadi** [comment]
@amcasey Ok.  Test added! :)

**Comment by gafter** [comment]
LGTM
